### PR TITLE
Refs [TASK][28356] Write abstract room database

### PR DESCRIPTION
**Pull request by aug16vcc**
## Related Tickets
- [#Ticket ID](https://edu-redmine.sun-asterisk.vn/issues/28356)

## WHAT
- Write abstract room database

## HOW
- Write abstract room database

## WHY
- First room class

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Vu Chi Cong | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/thooi/data/source/entity/User.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.sunasterisk.thooi.data.source.entity
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity
+data class User(
+    @PrimaryKey
+    val id: String
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/thooi/data/source/entity/User.kt:9]: sao có 1 thuộc tính thôi à ???

File: app/src/main/java/com/sunasterisk/thooi/data/source/local/database/AppDataBase.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package com.sunasterisk.thooi.data.source.local.database
+
+import android.content.Context
+import androidx.room.Database
+import androidx.room.Room
+import androidx.room.RoomDatabase
+import com.sunasterisk.thooi.data.source.entity.User
+
+@Database(entities = [User::class], version = DatabaseConstants.DATABASE_VERSION, exportSchema = false)
+abstract class AppDataBase : RoomDatabase() {
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/thooi/data/source/local/database/AppDataBase.kt:10]: liệu tên này chung chung quá k nhỉ, nếu có 2 database thì sao . ô có thể k cần sửa nhé

File: app/src/main/java/com/sunasterisk/thooi/data/source/local/database/AppDataBase.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        private var instance: AppDataBase? = null
+
+        fun getInstance(context: Context) = instance ?: synchronized(this) {
+            instance ?: buildAppDataBase(context).also { instance = it }
+        }
+
+        private fun buildAppDataBase(context: Context) =
+            Room.databaseBuilder(context, AppDataBase::class.java, DatabaseConstants.DATABASE_NAME)
+                .build()
+    }
```
Comment (thonc-2135) [inline app/src/main/java/com/sunasterisk/thooi/data/source/local/database/AppDataBase.kt]:
```suggestion
        fun getDatabase(context: Context): AppDataBase{
            return INSTANCE ?: synchronized(this) {
                Room.databaseBuilder(
                    context,
                    AppDataBase::class.java,
                   DatabaseConstants.DATABASE_NAME
                )
                    .fallbackToDestructiveMigration()
                    .build()
                    .also { INSTANCE = it }
            }
        }
```

File: app/src/main/java/com/sunasterisk/thooi/data/source/entity/User.kt (hunk 1 above)
Comment (aug16vcc) [reply app/src/main/java/com/sunasterisk/thooi/data/source/entity/User.kt:9]: Để không bị lỗi khi build ô ạ. Database cần ít nhất 1 entity nhưng viết ở task sau

File: app/src/main/java/com/sunasterisk/thooi/data/source/local/database/AppDataBase.kt (hunk 2 above)
Comment (aug16vcc) [reply app/src/main/java/com/sunasterisk/thooi/data/source/local/database/AppDataBase.kt:10]: Hiện tại thì để 1 database ô ạ

File: app/src/main/java/com/sunasterisk/thooi/data/source/entity/User.kt (hunk 1 above)
Comment (tuannm-1276) [reply app/src/main/java/com/sunasterisk/thooi/data/source/entity/User.kt:9]: lần sau comment vào nhé

**Pull request merged**